urllib3/transport_security.py

```diff
@@ -46,30 +57,112 @@ def validate_hpkp(self, conn):
             connected state.
         """
 
-    def process_response(self, response):
+    def process_response(self, response, conn):
         """
         Enroll or update hosts in our TSS based on response HSTS/HPKP
         headers.
-        This is a stub, to be implemented later.
 
         :param response:
             A :class:`urllib3.response.HTTPResponse` instance.
+        :param conn:
+            A :class:`urllib3.connection.HTTPSConnection` instance over which the response was
+            received.
         """
+        if isinstance(conn, HTTPSConnection):
+            sts_header = response.headers.get("strict-transport-security")
+            if sts_header:
+                self._process_sts_header(sts_header, conn.host)
+
+    def _process_sts_header(self, sts_header, host):
+        try:
+            parsed_sts_header = parse_header(sts_header)
+            sts_directives = dict((k.lower(), v) for k, v in parsed_sts_header)
+            if len(sts_directives) < len(parsed_sts_header):
+                raise HSTSError("Repeating directives in HSTS header, ignoring")
+            max_age = sts_directives.get("max-age", "")
+            if not max_age.isdigit():
+                raise HSTSError("Missing or invalid HSTS max-age directive, ignoring")
+            if is_ipaddress(host):
+                raise HSTSError("HSTS requested for an IP address, ignoring")
+            if "." not in host:
+                raise HSTSError("HSTS requested for a non-dot-separated name, ignoring")
+            if max_age == "0":
+                self._tss.invalidate_host(host)
+            else:
+                self._tss.store_host(host, force_https=True, max_age=max_age,
+                                     include_subdomains="includesubdomains" in sts_directives)
+        except Exception as e:
+            log.debug(e)
 
 
 class TransportSecurityStore(object):
     """
-    Abstract baseclass to store transport security (HSTS/HPKP) records.
+    Provides in-memory storage for transport security (HSTS/HPKP) records.
     """
+    MAX_AGE_LIMIT = 7776000  # 90 days; see https://tools.ietf.org/html/rfc6797#section-11.2
+
+    def __init__(self):
+        self._store = defaultdict(dict)
+
     def store_host(self, host, pins=None, force_https=False, include_subdomains=False,
                    max_age=None):
-        raise NotImplementedError("Must be overridden.")
+        max_age = min(int(max_age), self.MAX_AGE_LIMIT)
+        expires = int(time.time()) + max_age
+        self._store[host].update(force_https=force_https, include_subdomains=include_subdomains,
+                                 expires=expires)
 
     def invalidate_host(self, host):
-        raise NotImplementedError("Must be overridden.")
+        del self._store[host]
```

Comment:
I think this isn't right. Firstly, `invalidate_host` right now is called when HSTS sets `max-age=0`. That's fine, but when the host record is extended to include HPKP records this won't be right: simply turning off HSTS is not a reason to remove HPKP. Secondly, it makes `force_https` a redundant field: it can only ever be `True`, because `False` is equivalent to "there is no record".

That means I think we need two changes. Firstly, we need to be able to distinguish "there is no record" from "there is a record with HSTS not enabled". That's required for HPKP, and since we know it's coming we should build it in here.

Secondly, we should add newer methods for switching the status of a host. `store_host` should be able to store only the relevant fields for a given property (HSTS or HPKP).